EventFilter/RctRawToDigi/plugins/RctRawToDigi.cc

```diff
@@ -0,0 +1,381 @@
+    if ((int) rctRcd.size() < slinkHeaderSize_ + slinkTrailerSize_ + amc13HeaderSize_ + amc13TrailerSize_ ) {
+      LogError("L1T") << "Cannot unpack: empty/invalid L1T raw data (size = "
+		      << rctRcd.size() << ") for ID " << fedId_ << ". Returning empty collections!";
+      //continue;
+      return;
+    }
+
+    // do the unpacking
+    unpack(rctRcd, iEvent, colls.get()); 
+
+    // dump summary in verbose mode
+    if(verbose_) { doVerboseOutput(colls.get()); }
+    
+  }
+  else{
+
+    LogError("L1T") << "Cannot unpack: no collection found";
+
+    return; 
+  }
+
+}
+
+void RctRawToDigi::unpack(const FEDRawData& d, edm::Event& e, RctUnpackCollections * const colls)
+{
+
+  const unsigned char * data = d.data();  // The 8-bit wide raw-data array.  
+  // Data offset - starts at 16 as there is a 64-bit S-Link header followed
+  // by a 64-bit software-controlled header (for pipeline format version
+  // info that is not yet used).
+
+  
+  FEDHeader header(data);
+  
+  if (header.check()) {
+    LogDebug("L1T") << "Found SLink header:"
+		    << " Trigger type " << header.triggerType()
+		    << " L1 event ID " << header.lvl1ID()
+		    << " BX Number " << header.bxID()
+		    << " FED source " << header.sourceID()
+		    << " FED version " << header.version();
+  } else {
+    LogWarning("L1T") << "Did not find a valid SLink header!";
+  }
+  
+  FEDTrailer trailer(data + (d.size() - slinkTrailerSize_));
+  
+  if (trailer.check()) {
+    LogDebug("L1T") << "Found SLink trailer:"
+		    << " Length " << trailer.lenght()
+		    << " CRC " << trailer.crc()
+		    << " Status " << trailer.evtStatus()
+		    << " Throttling bits " << trailer.ttsBits();
+  } else {
+    LogWarning("L1T") << "Did not find a SLink trailer!";
+  }
+  
+  //amc13::Packet packet;
+  //if (!packet.parse((const uint64_t*) d.data(),d.size())) {
+  //LogError("L1T") << "Could not extract AMC13 Packet.";
+  //return;
+  //}
+    
+  size_t theSize = d.size();
+  uint32_t ch[(theSize+sizeof(uint32_t)-1)/sizeof(uint32_t)];
+  memcpy(ch,data,theSize);
+  unpackCTP7(ch, 0, sizeof(data), colls);
+
+}
+
+
+
+
+void RctRawToDigi::doVerboseOutput(const RctUnpackCollections * const colls) const
+{
+
+}
+
+
+
+void
+RctRawToDigi::unpackCTP7(const uint32_t *data, const unsigned block_id, const unsigned size, RctUnpackCollections * const colls)
+{
+  uint32_t of = 6;
+  LogDebug("L1T") << "Block ID  = " << block_id << " size = " << size;
+
+  CTP7Format ctp7Format;
+  RctDataDecoder rctDataDecoder;
+  uint32_t nBX = 0; 
+  uint32_t ctp7FWVersion;
+  uint32_t L1ID, L1aBCID;
+  std::vector<RCTInfo> allCrateRCTInfo[5];
+  
+  L1ID          =  data[1+of];                      // extract the L1 ID number
+  L1aBCID       =  data[5+of] & 0x00000FFF;         // extract the BCID number of L1A
+  nBX           = (data[5+of] & 0x00FF0000) >> 16;  // extract number of BXs readout per L1A 
+  ctp7FWVersion =  data[4+of];
+  
+  if(nBX != 1 && nBX != 3 && nBX != 5)
+    nBX = 1;
```

Comment:
OK, changed to nBXTemp. (The more fancy solution is neat but I do prefer 
the readability of using nBXTemp)

Thanks! Anything else?

On 6/29/15 9:23 PM, Chris Jones wrote:
>
> In EventFilter/RctRawToDigi/plugins/RctRawToDigi.cc 
> <https://github.com/cms-sw/cmssw/pull/9931#discussion_r33501590>:
>
> > +  LogDebug("L1T") << "Block ID  = " << block_id << " size = " << size;
> > +
> > +  CTP7Format ctp7Format;
> > +  RctDataDecoder rctDataDecoder;
> > +  uint32_t nBX = 0;
> > +  uint32_t ctp7FWVersion;
> > +  uint32_t L1ID, L1aBCID;
> > +  std::vector<RCTInfo> allCrateRCTInfo[5];
> > +
> > +  L1ID          =  data[1+of];                      // extract the L1 ID number
> > +  L1aBCID       =  data[5+of] & 0x00000FFF;         // extract the BCID number of L1A
> > +  nBX           = (data[5+of] & 0x00FF0000) >> 16;  // extract number of BXs readout per L1A
> > +  ctp7FWVersion =  data[4+of];
> > +
> > +  if(nBX != 1 && nBX != 3 && nBX != 5)
> > +    nBX = 1;
>
> It would be possible to declare nBX as |const| by doing
>
> |const uint32_t nBX = [] (uint32_t iData ) -> uint32_t {
>          uint32_t nBX = (data[5+of] & 0x00FF0000) >> 16;
>          if(nBX != 1 && nBX != 3 && nBX != 5) {
>              return 1;
>          }
>          return nBX;
>      }( data[5+of] );
> |
>
> Although I fear it isn't as readable. Alternatively you could declare 
> a |uint32_t nBXTemp| and do the calculation with that one and then do
>
> |     const uint32_t nBX = nBXTemp;
> |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/9931/files#r33501590>.
>